### ISSUE TITLE
blt deploy:build with existing project gives errors: "extras.installer-paths is missing from your composer.json file." and "sites/basebuildacquia/deploy/docroot/core" directory does not exist."

### ISSUE THREAD

**Issue by therobyouknow** [acquia/blt#2321]
I followed the steps here: https://blt.readthedocs.io/en/latest/readme/adding-to-project/

My system information:
* Operating system type: Mac OS X
* Operating system version: 10.12.6
* BLT version: BLT 8.9.10

Output of `blt doctor`:
```
lt-15-25:basebuildacquia robdavis$ blt doctor
Drupal VM was detected. Running blt doctor inside of VM...
Executing command cd /var/www/basebuildacquia/docroot/.. && /var/www/basebuildacquia/docroot/../vendor/bin/drush cc drush && /var/www/basebuildacquia/docroot/../vendor/bin/drush --include=/var/www/basebuildacquia/docroot/../vendor/acquia/blt/drush blt-doctor -r /var/www/basebuildacquia/docroot/../docroot inside of Drupal VM...
bash: line 0: cd: /var/www/basebuildacquia: No such file or directory
bash: line 0: cd: /var/www/basebuildacquia/docroot/..: No such file or directory
Connection to 127.0.0.1 closed.
Attempting to run doctor on host machine...
[warning] Unable to run the doctor using alias '@basebuildacquia.local'. Trying with '@self'...
[warning] Unable to run the doctor using alias '@self'. Trying without alias...
[error]  Unable to execute the `blt doctor` command. 
lt-15-25:basebuildacquia robdavis$ 
```

When I run this command:
```
blt deploy:build
```

I get the following output:
```
lt-15-25:basebuildacquia robdavis$ blt deploy:build
Generating build artifact...
For more detailed output, use the -v flag.
> frontend
Executing command blt frontend --define=drush.alias=self inside of Drupal VM...
bash: line 0: cd: /var/www/basebuildacquia: No such file or directory
bash: blt: command not found
Connection to 127.0.0.1 closed.
> setup:hash-salt
Hash salt already exists.
Combining exclusions from deploy.deploy-exclude-additions and deploy.deploy-exclude files...
Rsyncing files from source repo into the build artifact...
Rebuilding composer dependencies for production...
[ExecStack] composer install --no-dev --no-interaction --optimize-autoloader
[ExecStack] Running composer install --no-dev --no-interaction --optimize-autoloader in /Users/robdavis/sites/basebuildacquia/deploy
Loading composer repositories with package information
Installing dependencies from lock file
Package operations: 65 installs, 0 updates, 0 removals
  - Installing wikimedia/composer-merge-plugin (v1.4.1): Loading from cache
  - Installing composer/installers (v1.4.0): Loading from cache
  - Installing grasmash/drupal-security-warning (1.1.1): Loading from cache
  - Installing tivie/php-os-detector (1.1.0): Loading from cache
  - Installing symfony/yaml (v3.2.14): Loading from cache
  - Installing symfony/polyfill-mbstring (v1.6.0): Loading from cache
  - Installing psr/log (1.0.2): Loading from cache
  - Installing symfony/debug (v3.3.13): Loading from cache
  - Installing symfony/console (v3.2.14): Loading from cache
  - Installing squizlabs/php_codesniffer (2.9.1): Loading from cache
  - Installing sebastian/version (1.0.6): Loading from cache
  - Installing sebastian/global-state (1.1.1): Loading from cache
  - Installing sebastian/recursion-context (1.0.5): Loading from cache
  - Installing sebastian/exporter (1.2.2): Loading from cache
  - Installing sebastian/environment (1.3.8): Loading from cache
  - Installing sebastian/diff (1.4.3): Loading from cache
  - Installing sebastian/comparator (1.2.4): Loading from cache
  - Installing phpunit/php-text-template (1.2.1): Loading from cache
  - Installing doctrine/instantiator (1.0.5): Loading from cache
  - Installing phpunit/phpunit-mock-objects (2.3.8): Loading from cache
  - Installing phpunit/php-timer (1.0.9): Loading from cache
  - Installing phpunit/php-file-iterator (1.4.5): Loading from cache
  - Installing phpunit/php-token-stream (1.4.11): Loading from cache
  - Installing phpunit/php-code-coverage (2.2.4): Loading from cache
  - Installing phpdocumentor/reflection-docblock (2.0.5): Loading from cache
  - Installing phpspec/prophecy (1.7.3): Loading from cache
  - Installing phpunit/phpunit (4.8.36): Loading from cache
  - Installing oomphinc/composer-installers-extender (v1.1.2): Loading from cache
  - Installing dflydev/dot-access-data (v1.1.0): Loading from cache
  - Installing grasmash/yaml-expander (1.2.0): Loading from cache
  - Installing symfony/filesystem (v3.3.13): Loading from cache
  - Installing grasmash/yaml-cli (1.0.3): Loading from cache
  - Installing webmozart/assert (1.2.0): Loading from cache
  - Installing webmozart/path-util (2.3.0): Loading from cache
  - Installing symfony/var-dumper (v3.3.13): Loading from cache
  - Installing symfony/finder (v3.3.13): Loading from cache
  - Installing symfony/event-dispatcher (v3.3.13): Loading from cache
  - Installing nikic/php-parser (v3.1.2): Loading from cache
  - Installing jakub-onderka/php-console-color (0.1): Loading from cache
  - Installing jakub-onderka/php-console-highlighter (v0.3.2): Loading from cache
  - Installing dnoegel/php-xdg-base-dir (0.1): Loading from cache
  - Installing psy/psysh (v0.8.15): Loading from cache
  - Installing pear/console_table (v1.3.0): Loading from cache
  - Installing consolidation/output-formatters (3.1.13): Loading from cache
  - Installing consolidation/annotated-command (2.8.2): Loading from cache
  - Installing drush/drush (8.1.15): Loading from cache
  - Installing drupal/coder (8.2.12): Cloning 984c54a7b1 from cache
  - Installing doctrine/inflector (v1.1.0): Loading from cache
  - Installing doctrine/lexer (v1.0.1): Loading from cache
  - Installing doctrine/collections (v1.4.0): Loading from cache
  - Installing doctrine/cache (v1.6.2): Loading from cache
  - Installing doctrine/annotations (v1.4.0): Loading from cache
  - Installing doctrine/common (v2.7.3): Loading from cache
  - Installing dealerdirect/phpcodesniffer-composer-installer (v0.4.3): Loading from cache
  - Installing symfony/process (v3.3.13): Loading from cache
  - Installing psr/container (1.0.0): Loading from cache
  - Installing container-interop/container-interop (1.2.0): Loading from cache
  - Installing league/container (2.4.1): Loading from cache
  - Installing consolidation/log (1.0.5): Loading from cache
  - Installing consolidation/config (1.0.7): Loading from cache
  - Installing consolidation/robo (1.1.5): Loading from cache
  - Installing composer/semver (1.4.2): Loading from cache
  - Installing twig/twig (v1.35.0): Loading from cache
  - Installing asm89/twig-lint (1.0.2): Loading from cache
  - Installing acquia/blt (8.9.10): Loading from cache
Generating optimized autoload files
Error: extras.installer-paths is missing from your composer.json file.
Loading composer repositories with package information
Installing dependencies from lock file
Nothing to install or update
Generating optimized autoload files
PHP CodeSniffer Config installed_paths set to ../../drupal/coder/coder_sniffer/
PHP CodeSniffer Config installed_paths set to ../../drupal/coder/coder_sniffer/
[ExecStack] Done in 19.658s
Sanitizing artifact...
[error]  The "/Users/robdavis/sites/basebuildacquia/deploy/docroot/core" directory does not exist. 
```

And I expected this to happen:
- Something to indicate that it was successful

**Comment by therobyouknow** [comment]
I added the following to my composer.json

```
        "installer-paths": {
            "docroot/core": ["type:drupal-core"],
            "docroot/modules/contrib/{$name}": ["type:drupal-module"],
            "docroot/modules/custom/{$name}": ["type:drupal-custom-module"],
            "docroot/profiles/contrib/{$name}": ["type:drupal-profile"],
            "docroot/profiles/custom/{$name}": ["type:drupal-custom-profile"],
            "docroot/themes/contrib/{$name}": ["type:drupal-theme"],
            "docroot/themes/custom/{$name}": ["type:drupal-custom-theme"],
            "docroot/libraries/{$name}": ["type:drupal-library", "type:bower-asset", "type:npm-asset"],
            "drush/contrib/{$name}": ["type:drupal-drush"]
```

as per

https://github.com/acquia/blt/blob/8.x/template/composer.json

Which fixed the error "extras.installer-paths is missing from your composer.json file." - misleadingly, the error refers to extras (plural) but actually within the json file the word extra has to be used (singular).

I don't yet know how to solve the 2nd error: ""sites/basebuildacquia/deploy/docroot/core" directory does not exist." - I tried creating this folder but its parent gets overwritten when I run the blt deploy:build command.

**Comment by therobyouknow** [comment]
I think I've solved the second error

In my docroot I did:

` ln -s html/docroot .`

because docroot is one folder down in html in my project that I am trying to get to work with BLT.

I now see:
```
The deployment artifact was generated at /Users/robdavis/sites/basebuildacquia/deploy.
lt-15-25:basebuildacquia robdavis$ history
```

...before the blt deploy:build command completes, which indicates it was successful.

**Comment by danepowell** [comment]
Like #2320 I suspect this will be resolved in 9.1.x, however since it sounds like you've made some progress in fixing this feel free to open a PR for 8.9.x. That template composer.json _should_ get merged with your existing project composer.json in order to add the installer paths when a new project is created, not sure why that's not happening. It's probably dependent on your particular project structure.

**Comment by grasmash** [comment]
I'm concerned that you may be using an abnormal directory structure with BLT. BLT requires that you use a specific directory structure.

You should have, at a minimum, the following structure:

box
composer.json
docroot

Do you have those files and directories? Also, what process did you use to create your project?
